src/payment/bolt12.rs

```diff
@@ -0,0 +1,272 @@
+				return Err(Error::InvalidOffer);
+			},
+		};
+
+		match self.channel_manager.pay_for_offer(
+			&offer,
+			quantity,
+			amount_msats,
+			payer_note,
+			payment_id,
+			retry_strategy,
+			max_total_routing_fee_msat,
+		) {
+			Ok(()) => {
+				let payee_pubkey = offer.signing_pubkey();
+				log_info!(
+					self.logger,
+					"Initiated sending {:?} to {:?}",
+					offer_amount_msat,
+					payee_pubkey
+				);
+
+				let kind = PaymentKind::Bolt12 { hash: None, preimage: None, secret: None };
+				let payment = PaymentDetails {
+					id: payment_id,
+					kind,
+					amount_msat: Some(*offer_amount_msat),
+					direction: PaymentDirection::Outbound,
+					status: PaymentStatus::Pending,
+				};
+				self.payment_store.insert(payment)?;
+
+				Ok(payment_id)
+			},
+			Err(e) => {
+				log_error!(self.logger, "Failed to send payment: {:?}", e);
+				match e {
+					Bolt12SemanticError::DuplicatePaymentId => Err(Error::DuplicatePayment),
+					_ => {
+						let kind = PaymentKind::Bolt12 { hash: None, preimage: None, secret: None };
+						let payment = PaymentDetails {
+							id: payment_id,
+							kind,
+							amount_msat: Some(*offer_amount_msat),
+							direction: PaymentDirection::Outbound,
+							status: PaymentStatus::Failed,
+						};
+						self.payment_store.insert(payment)?;
+						Err(Error::PaymentSendingFailed)
+					},
+				}
+			},
+		}
+	}
+
+	/// Send a payment given an offer and an amount in millisatoshi.
+	///
+	/// This will fail if the amount given is less than the value required by the given offer.
+	///
+	/// This can be used to pay a so-called "zero-amount" offers, i.e., an offer that leaves the
+	/// amount paid to be determined by the user.
+	///
+	/// If `payer_note` is `Some` it will be seen by the recipient and reflected back in the invoice
+	/// response.
+	pub fn send_using_amount(
+		&self, offer: &Offer, payer_note: Option<String>, amount_msat: u64,
+	) -> Result<PaymentId, Error> {
+		let rt_lock = self.runtime.read().unwrap();
+		if rt_lock.is_none() {
+			return Err(Error::NotRunning);
+		}
+
+		let quantity = None;
+		let amount_msats = None;
+		let mut random_bytes = [0u8; 32];
+		rand::thread_rng().fill_bytes(&mut random_bytes);
+		let payment_id = PaymentId(random_bytes);
+		let retry_strategy = Retry::Timeout(LDK_PAYMENT_RETRY_TIMEOUT);
+		let max_total_routing_fee_msat = None;
+
+		let offer_amount_msat = match offer.amount() {
+			Some(Amount::Bitcoin { amount_msats }) => *amount_msats,
+			Some(_) => {
+				log_error!(self.logger, "Failed to send payment as the provided offer was denominated in an unsupported currency.");
+				return Err(Error::UnsupportedCurrency);
+			},
+			None => amount_msat,
+		};
+
+		if amount_msat < offer_amount_msat {
+			log_error!(
+				self.logger,
+				"Failed to pay as the given amount needs to be at least the offer amount: required {}msat, gave {}msat.", offer_amount_msat, amount_msat);
+			return Err(Error::InvalidAmount);
+		}
```

Comment:
Maybe do this inside the above match arm so that you don't later shadow `offer_amount_msat`.